File: leetcode/javascript/0012_integer_to_roman.js
// https://leetcode.com/problems/integer-to-roman/
// O(log(n) + k) time complexity
// O(k) space complexity, where
// n is the input number and k is the length of converted roman numeral
function intToRoman(num) {
  const MAP_INT_TO_ROMAN = new Map([
    [1000, ['_', '_', 'M']],
    [100, ['M', 'D', 'C']],
    [10, ['C', 'L', 'X']],
    [1, ['X', 'V', 'I']],
  ]);

  const romanConverterHelper = (arr, num, factorOfTen, romanNumerals) => {
    while (num >= factorOfTen) {
      if (num >= 9 * factorOfTen) {
        arr.push(romanNumerals[2] + romanNumerals[0]);
        num -= 9 * factorOfTen;
      } else if (num >= 5 * factorOfTen) {
        arr.push(romanNumerals[1]);
        num -= 5 * factorOfTen;
      } else if (num >= 4 * factorOfTen) {
        arr.push(romanNumerals[2] + romanNumerals[1]);
        num -= 4 * factorOfTen;
      } else {
        arr.push(romanNumerals[2]);
        num -= factorOfTen;
      }
    }
    return num;
  };
  let convertedToRoman = [];
  for (const [factorOfTen, romanNumerals] of MAP_INT_TO_ROMAN.entries()) {
    num = romanConverterHelper(
      convertedToRoman,
      num,
      factorOfTen,
      romanNumerals,
    );
  }
  return convertedToRoman.join('');
}

File: leetcode/python/0012_integer_to_roman.py
# https://leetcode.com/problems/integer-to-roman/
# O(1) time complexity because numbers constrained to between 1 and 4000
# O(1) space complexity
def integer_to_roman(num: int) -> str:
    NUMERAL_MAPPING = {
        1000: ('_', '_', 'M'),
        100: ('M', 'D', 'C'),
        10: ('C', 'L', 'X'),
        1: ('X', 'V', 'I')
    }

    list_of_numerals = []
    for factor in [1000, 100, 10, 1]:
        numerals = NUMERAL_MAPPING[factor]
        while num >= factor:
            if num >= 9 * factor:
                list_of_numerals.append(numerals[2] + numerals[0])
                num -= 9 * factor
            elif num >= 5 * factor:
                list_of_numerals.append(numerals[1])
                num -= 5 * factor
            elif num >= 4 * factor:
                list_of_numerals.append(numerals[2] + numerals[1])
                num -= 4 * factor
            else:
                list_of_numerals.append(numerals[2])
                num -= factor

    return ''.join(list_of_numerals)
